hal/src/thumbv7em/adc.rs

```diff
@@ -235,100 +297,111 @@ impl<C> InterruptAdc<$ADC, C>
     }
 }
 
-impl<C> From<Adc<$ADC>> for InterruptAdc<$ADC, C>
-    where C: ConversionMode<$ADC>
+impl<ADC, C> From<Adc<ADC>> for InterruptAdc<ADC, C>
+where
+    ADC: AdcPeripheral,
+    C: ConversionMode<ADC>,
 {
-    fn from(adc: Adc<$ADC>) -> Self {
+    fn from(adc: Adc<ADC>) -> Self {
         Self {
             adc,
-            m: core::marker::PhantomData{},
+            m: core::marker::PhantomData {},
         }
     }
 }
 
-impl<WORD, PIN> OneShot<$ADC, WORD, PIN> for Adc<$ADC>
+impl<ADC, WORD, PIN> OneShot<ADC, WORD, PIN> for Adc<ADC>
 where
-   WORD: From<u16>,
-   PIN: Channel<$ADC, ID=u8>,
+    ADC: AdcPeripheral,
+    WORD: From<u16>,
+    PIN: Channel<ADC, ID = u8>,
 {
-   type Error = ();
+    type Error = ();
 
-   fn read(&mut self, pin: &mut PIN) -> nb::Result<WORD, Self::Error> {
+    fn read(&mut self, pin: &mut PIN) -> nb::Result<WORD, Self::Error> {
         self.mux(pin);
         self.power_up();
         let result = self.synchronous_convert();
         self.power_down();
         Ok(result.into())
-   }
-}
-        )+
     }
 }
 
-adc_hal! {
-    ADC0: (adc0, apbdmask, adc0_, adc0_biascomp_scale_cal, adc0_biasref_scale_cal, adc0_biasr2r_scale_cal),
-    ADC1: (adc1, apbdmask, adc1_, adc1_biascomp_scale_cal, adc1_biasref_scale_cal, adc1_biasr2r_scale_cal),
+pub trait AdcChannel<ADC: AdcPeripheral>: PinId {
+    const CHANNEL: u8;
+}
+
+impl<ADC, I> Channel<ADC> for Pin<I, AlternateB>
+where
+    ADC: AdcPeripheral,
+    I: AdcChannel<ADC>,
+{
+    type ID = u8;
+    fn channel() -> u8 {
+        I::CHANNEL
+    }
 }
 
 macro_rules! adc_pins {
     (
         $(
-            $PinId:ident: ($ADC:ident, $CHAN:literal),
+            $PinId:ident: [ $( ($ADC:ident, $CHAN:literal) ),+ ],
```

Comment:
I made this a list of tuples rather than a single tuple, because some pins can be an input for both ADC0 and ADC1. We should update the pin table below to match the datasheet. It's missing a bunch of entries right now.